test/storage-luatest/storage_1_1_test.lua

```diff
@@ -157,3 +157,76 @@ test_group.test_bucket_send_field_types = function(g)
         box.space.test:drop()
     end)
 end
+
+--
+-- Custom triggers on bucket changes
+--
+test_group.test_on_bucket_event = function(g)
+    -- Initial problem: buckets migration looks like a regular
+    -- client data activity, indistinguishable in the replication flow.
+    -- In order to allow efficient change data capturing we can add a fake
+    -- operation in the beginning of the transaction, so that a client could
+    -- determing which operations can be skipped (using HEADER.tsn) as they
+    -- were done not by the user but by gc or rebalancer.
+    vtest.cluster_exec_each_master(g, function(engine)
+        -- Space, in which these fake operations will be written
+        rawset(_G, 'notify_name', 'test_migration_notify_' .. engine)
+        box.schema.space.create(_G.notify_name, {
+            engine = engine,
+        }):create_index('primary')
+        local s = box.schema.space.create('data', {
+             engine = engine,
+             format = {
+                {'id', 'unsigned'},
+                {'bucket_id', 'unsigned'},
+             },
+        })
+        s:create_index('id', {parts = {'id'}})
+        s:create_index('bucket_id', {parts = {'bucket_id'}, unique = false})
+
+        -- Set on_bucket_event trigger
+        rawset(_G, 'migration_notifications', {})
+        rawset(_G, 'notify_trigger', function(event_type, bucket_id, opts)
+            -- Set on_commit trigger
+            box.on_commit(function(iterator)
+                local notify_space = box.space[_G.notify_name]
+                for idx, _, _, space_id in iterator() do
+                    if space_id == notify_space.id then
+                        table.insert(_G.migration_notifications, idx)
+                    end
+                end
+            end)
+            ilt.assert_not_equals(opts and opts.spaces, nil)
+            if event_type == 'bucket_data_recv_txn' or
+                    event_type == 'bucket_data_gc_txn' then
+                box.space[_G.notify_name]:delete({bucket_id})
+            end
+        end)
+        ivshard.storage.on_bucket_event(_G.notify_trigger)
+    end, {g.params.engine})
+
+    g.replica_2_a:exec(function()
+        local bid = _G.get_first_bucket()
+        local uuid = ivtest.replicaset_name_to_uuid('replica_1_a')
+        box.space.data:insert({bid, bid})
+        ivshard.storage.bucket_send(bid, uuid)
+        _G.bucket_gc_wait()
+    end)
+
+    -- Rebalancer related notifications
+    g.replica_1_a:exec(function()
+        ilt.assert_equals(#_G.migration_notifications, 1)
+    end)
+
+    -- GC related notifications
+    g.replica_2_a:exec(function()
+        ilt.assert_equals(#_G.migration_notifications, 1)
+    end)
+
+    -- Cleanup
+    vtest.cluster_exec_each_master(g, function()
+        ivshard.storage.on_bucket_event(nil, _G.notify_trigger)
+        box.space[_G.notify_name]:drop()
+        box.space.data:drop()
+    end)
+end
```

Comment:
Please, apply this diff:
<details><summary>Diff</summary>

```Diff
diff --git a/test/storage-luatest/storage_1_1_test.lua b/test/storage-luatest/storage_1_1_test.lua
index e1b6229..cd49050 100644
--- a/test/storage-luatest/storage_1_1_test.lua
+++ b/test/storage-luatest/storage_1_1_test.lua
@@ -1,4 +1,5 @@
 local t = require('luatest')
+local vconst = require('vshard.consts')
 local vtest = require('test.luatest_helpers.vtest')
 local vutil = require('vshard.util')
 
@@ -159,74 +160,107 @@ test_group.test_bucket_send_field_types = function(g)
 end
 
 --
--- Custom triggers on bucket changes
+-- Custom triggers on bucket changes.
 --
 test_group.test_on_bucket_event = function(g)
-    -- Initial problem: buckets migration looks like a regular
-    -- client data activity, indistinguishable in the replication flow.
-    -- In order to allow efficient change data capturing we can add a fake
-    -- operation in the beginning of the transaction, so that a client could
-    -- determing which operations can be skipped (using HEADER.tsn) as they
-    -- were done not by the user but by gc or rebalancer.
     vtest.cluster_exec_each_master(g, function(engine)
-        -- Space, in which these fake operations will be written
-        rawset(_G, 'notify_name', 'test_migration_notify_' .. engine)
-        box.schema.space.create(_G.notify_name, {
+        local event_space = box.schema.space.create('events', {
             engine = engine,
-        }):create_index('primary')
-        local s = box.schema.space.create('data', {
-             engine = engine,
-             format = {
-                {'id', 'unsigned'},
-                {'bucket_id', 'unsigned'},
-             },
         })
-        s:create_index('id', {parts = {'id'}})
-        s:create_index('bucket_id', {parts = {'bucket_id'}, unique = false})
+        event_space:create_index('primary')
+
+        local function data_space_new(name)
+            local s = box.schema.space.create(name, {
+                engine = engine,
+                format = {
+                    {'id', 'unsigned'},
+                    {'bucket_id', 'unsigned'},
+                },
+            })
+            s:create_index('id', {parts = {'id'}})
+            s:create_index('bucket_id', {
+                parts = {'bucket_id'}, unique = false
+            })
+            return s
+        end
+        local data_space1 = data_space_new('data1')
+        local data_space2 = data_space_new('data2')
+
+        local test_txn_id = 0
+        local test_events = {}
+        local test_commits = {}
+        rawset(_G, 'test_trigger', function(event_type, bucket_id, data)
+            test_txn_id = test_txn_id + 1
+            local tid = test_txn_id
+            local event_data = {event_type, bucket_id, data}
+            ilt.assert(box.is_in_txn())
+            ilt.assert(not test_events[tid])
+            event_space:insert{tid, event_data}
+            test_events[tid] = event_data
 
-        -- Set on_bucket_event trigger
-        rawset(_G, 'migration_notifications', {})
-        rawset(_G, 'notify_trigger', function(event_type, bucket_id, opts)
-            -- Set on_commit trigger
             box.on_commit(function(iterator)
-                local notify_space = box.space[_G.notify_name]
-                for idx, _, _, space_id in iterator() do
-                    if space_id == notify_space.id then
-                        table.insert(_G.migration_notifications, idx)
+                for _, t1, t2, space_id in iterator() do
+                    if space_id == event_space.id then
+                        test_commits[tid] = true
+                        break
                     end
                 end
             end)
-            ilt.assert_not_equals(opts and opts.spaces, nil)
-            if event_type == 'bucket_data_recv_txn' or
-                    event_type == 'bucket_data_gc_txn' then
-                box.space[_G.notify_name]:delete({bucket_id})
+        end)
+        ivshard.storage.on_bucket_event(_G.test_trigger)
+
+        rawset(_G, 'process_events', function(expected_events)
+            -- Compare only the items, not looking at the order. Because the
+            -- spaces can be sent in any order, can't rely on one.
+            ilt.assert_items_equals(test_events, expected_events)
+            -- Ensure that each saved event is also persisted.
+            for tid, event_data in pairs(test_events) do
+                ilt.assert_equals(box.space.events:get{tid}, {tid, event_data})
+                ilt.assert_not_equals(test_commits[tid], nil)
+                test_commits[tid] = nil
+                box.space.events:delete{tid}
             end
+            test_events = {}
+            -- No unexpected events anywhere.
+            ilt.assert(not next(test_commits))
+            ilt.assert_equals(box.space.events.index[0]:min(), nil)
         end)
-        ivshard.storage.on_bucket_event(_G.notify_trigger)
     end, {g.params.engine})
 
-    g.replica_2_a:exec(function()
+    local rs1_uuid = g.replica_1_a:exec(function()
+        return box.info.cluster.uuid
+    end)
+    local bid = g.replica_2_a:exec(function(rs1_uuid)
         local bid = _G.get_first_bucket()
-        local uuid = ivtest.replicaset_name_to_uuid('replica_1_a')
-        box.space.data:insert({bid, bid})
-        ivshard.storage.bucket_send(bid, uuid)
+        box.space.data1:insert({10, bid})
+        box.space.data2:insert({20, bid})
+        local ok, err = ivshard.storage.bucket_send(
+            bid, rs1_uuid, {timeout = _G.iwait_timeout})
+        ilt.assert_equals(err, nil)
+        ilt.assert(ok)
         _G.bucket_gc_wait()
-    end)
+        return bid
+    end, {rs1_uuid})
 
-    -- Rebalancer related notifications
-    g.replica_1_a:exec(function()
-        ilt.assert_equals(#_G.migration_notifications, 1)
-    end)
+    g.replica_2_a:exec(function(events)
+        _G.process_events(events)
+    end, {{
+        [1] = {vconst.BUCKET_TRIGGER.GC, bid, {spaces = {'data1'}}},
+        [2] = {vconst.BUCKET_TRIGGER.GC, bid, {spaces = {'data2'}}},
+    }})
 
-    -- GC related notifications
-    g.replica_2_a:exec(function()
-        ilt.assert_equals(#_G.migration_notifications, 1)
-    end)
+    g.replica_1_a:exec(function(events)
+        _G.process_events(events)
+    end, {{
+        [1] = {vconst.BUCKET_TRIGGER.RECV, bid, {spaces = {'data1'}}},
+        [2] = {vconst.BUCKET_TRIGGER.RECV, bid, {spaces = {'data2'}}},
+    }})
 
     -- Cleanup
     vtest.cluster_exec_each_master(g, function()
-        ivshard.storage.on_bucket_event(nil, _G.notify_trigger)
-        box.space[_G.notify_name]:drop()
-        box.space.data:drop()
+        ivshard.storage.on_bucket_event(nil, _G.test_trigger)
+        box.space.events:drop()
+        box.space.data1:drop()
+        box.space.data2:drop()
     end)
 end
diff --git a/vshard/consts.lua b/vshard/consts.lua
index f606666..5d6e058 100644
--- a/vshard/consts.lua
+++ b/vshard/consts.lua
@@ -10,8 +10,11 @@ return {
     },
 
     BUCKET_TRIGGER = {
-      RECV = 'bucket_data_recv_txn',
-      GC = 'bucket_data_gc_txn'
+        -- Txn triggers allow to attach anything to the bucket transaction. That
+        -- was requested by a customer to be able to detect vshard book-keeping
+        -- actions like bucket sending right in the journal.
+        RECV = 'bucket_data_recv_txn',
+        GC = 'bucket_data_gc_txn'
     },
 
     STATUS = {
```
</details>

The problems with the old test were that
- Its comment rather belongs to the event declarations;
- Need to test the third argument of the trigger. Testing just for `~= nil` wouldn't be enough. We could break the third argument layout and the test wouldn't notice;
- Need to test multiple spaces to ensure that they are split into separate txns and events;
- Need to test the other 2 arguments of the event as well. Testing just event count wouldn't be enough;
- I am not sure I understood this `box.space[_G.notify_name]:delete({bucket_id})` line. What is it for? It would be just NOP. Because this space was always empty in your test.
- Need to test if `bucket_send()` was ok. It wouldn't raise an exception on an error.
- About `ivtest.replicaset_name_to_uuid('replica_1_a')` I was wrong, sorry. It won't work properly in general case. It generates the UUIDs on fly. In your version of the test it worked by luck, because `replica_1_a` on the storage got the same UUID as `replicaset 1` on the test node. But otherwise it is obviously wrong - we tried to use **replicaset**_name_to_uuid to get UUID of a **replica**. Replicaset and replica UUIDs are not related.

Reply:
The new test looks great, applied it almost entirely. Just fixed places, luacheck doesn't like